web/pdf_history.js

```diff
@@ -13,21 +13,455 @@
+        // Providing the third argument causes a SecurityError for file:// URLs.
+        window.history.pushState(newState, '');
+      } else {
+        window.history.pushState(newState, '', document.URL);
+      }
+    }
+
+    if (typeof PDFJSDev !== 'undefined' && PDFJSDev.test('CHROME') &&
+        top === window) {
+      // eslint-disable-next-line no-undef
+      chrome.runtime.sendMessage('showPageAction');
+    }
+  }
+
+  /**
+   * @private
+   */
+  _tryPushCurrentPosition() {
+    if (!this._position) {
+      return;
+    }
+    let position = this._position;
+
+    if (!this._destination) {
+      this._pushOrReplaceState(position);
+      return;
+    }
+    if (this._destination.hash === position.hash) {
+      return; // The current document position has not changed.
+    }
+    if (!this._destination.page) {
+      // `this._destination` was set through the user changing the hash of
+      // the document. Do not add `this._position` to the browser history,
+      // to avoid "flooding" it with lots of (nearly) identical entries,
+      // since we cannot ensure that the document position has changed.
+      return;
+    }
+
+    let forceReplace = false;
+    if (this._destination.page === position.first ||
+        this._destination.page === position.page) {
+      // When the `page` of `this._destination` is still visible, do not
+      // update the browsing history when `this._destination` either:
+      //  - contains an internal destination, since in this case we
+      //    cannot ensure that the document position has actually changed.
+      //  - was set through the user changing the hash of the document.
+      if (this._destination.dest || !this._destination.first) {
+        return;
+      }
+      // To avoid "flooding" the browser history, replace the current entry.
+      forceReplace = true;
+    }
+    this._pushOrReplaceState(position, forceReplace);
+  }
+
+  /**
+   * @private
+   */
+  _isValidState(state) {
+    if (!state) {
+      return false;
+    }
+    if (state.fingerprint !== this.fingerprint) {
+      // This should only occur in viewers with support for opening more than
+      // one PDF document, e.g. the GENERIC viewer.
+      return false;
+    }
+    if (!Number.isInteger(state.uid) || state.uid < 0) {
+      return false;
+    }
+    if (typeof state.destination !== 'object' || state.destination === null) {
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * @private
+   */
+  _resetSomeInternalState(destination, uid) {
+    this._destination = destination;
+    this._currentUid = uid;
+    this._uid = this._currentUid + 1;
+  }
+
+  /**
+   * @private
+   */
+  _updateViewarea({ location, }) {
+    this._position = {
+      hash: this._isViewerInPresentationMode ?
+        `page=${location.pageNumber}` : location.pdfOpenParams.substring(1),
+      page: this.linkService.page,
+      first: location.pageNumber,
+    };
+
+    if (this._popStateInProgress) {
```

Comment:
This appears to have no effect since the method ends here anyway. Should this be above setting `this._position` instead?

Reply:
Nope, it should *not* be moved, but it will become relevant in the very next commit.